src/model/Model.js

```diff
@@ -0,0 +1,347 @@
+import {InvalidQuery, InvalidValue} from '../errors'
+import {Page, PageIterator, Ref} from '../objects'
+import * as query from '../query'
+import Field from './Field'
+import {calculateDiff, getPath, objectDup, setPath} from './_util'
+
+/**
+ * Base class for all models.
+ *
+ * Models represent database instances.
+ * They link a FaunaDB class to a JavaScript class.
+ *
+ * The basic format is:
+ *
+ *     class MyModel extends Model {
+ *       ... your methods ...
+ *     }
+ *     // define class name and fields
+ *     MyModel.setup('my_models', {
+ *       x: {},
+ *       y: {converter: new RefConverter(MyModel)}
+ *     })
+ *
+ * {@link Field}s will be constructed and
+ * properties will be generated for each field passed to {@link setup}.
+ *
+ * {@link Class.createForModel} must be called before you can save model instances.
+ */
+export default class Model {
+  /**
+   * @param {string} faunaClassName
+   * @param {object} fields
+   *   Each `key: value` pair is the parameters for `addField`.
+   */
+  static setup(faunaClassName, fields={}) {
+    this.faunaClassName = faunaClassName
+    /**
+     * {@link Ref} for the class itself.
+     *
+     * `instance.ref` should be the same as `new Ref(instance.constructor.classRef, instance.id)`.
+     */
+    this.classRef = new Ref('classes', faunaClassName)
+    /** Object of all fields assigned to this class. */
+    this.fields = {}
+    for (const fieldName in fields)
+      this.addField(fieldName, fields[fieldName])
+  }
+
+  /**
+   * Adds a new field to the class, making getters and setters.
+   *
+   * @param {string} fieldName
+   *   Name for the field. A getter and setter will be made with this name.
+   *   If `fieldOpts.path` is not defined, it defaults to `['data', fieldName]`.
+   * @param {object} fieldOpts
+   *   Parameters for the {@link Field} constructor.
+   */
+  static addField(fieldName, fieldOpts={}) {
+    if (fieldName === 'ref' || fieldName === 'ts')
+      throw new Error('Forbidden field name.')
+
+    if (fieldOpts.path == null)
+      fieldOpts.path = ['data', fieldName]
+    const field = new Field(fieldOpts)
+    this.fields[fieldName] = field
+
+    const {get, set} = field.codec === null ?
+      {
+        get() {
+          return getPath(field.path, this._current)
+        },
+        set(value) {
+          setPath(field.path, value, this._current)
+        }
+      } : {
+        get() {
+          if (fieldName in this._cache)
+            return this._cache[fieldName]
+          else {
+            const encoded = getPath(field.path, this._current)
+            const decoded = field.codec.decode(encoded, this)
+            this._cache[fieldName] = decoded
```

Comment:
are codecs expected to be expensive? If not, I'd say that we do away with the cache entirely and call the codec on get/set unconditionally